### ISSUE TITLE
Feature: Check referenced Action SHA to ensure authenticity

### ISSUE THREAD

**Issue by cedws** [woodruffw/zizmor#271]
### Pre-submission checks

- [X] I am **not** reporting a bug (crash, false positive/negative, etc). These must be filed via the bug report template.
- [X] I have looked through the open issues for a duplicate request.

### What's the problem this feature will solve?

Zizmor already supports checking Action versions used within in a workflow are pinned to a specific SHA, when ran in pedantic mode. IMO this is highly valuable to ensure supply chain security.

```
help[unpinned-uses]: unpinned action reference
  --> /Users/dev/Documents/Projects/iapc/.github/workflows/build.yml:16:9
   |
16 |         uses: actions/setup-go@v5
   |         ------------------------- help: action is not pinned to a hash ref
   |
   = note: audit confidence → High
```

An example correction for this warning would be to pin the tag to `@3041bf56c941b39c61721a86cd11f3bb1338122a`, ensuring the author of the Action can't 'rugpull' and compromise the workflow by moving the tag to a malicious revision.

However, due to the way GitHub stores objects, that SHA can actually be malicious. GitHub considers all objects in an upstream repo and forks to be part of the same 'namespace.' Therefore, the SHA can point to a forked version of an Action, while the actual Action usage looks authentic.

Although you'd hope that any malicious Action SHA changes would be caught at review time, many developers don't know that this is how GItHub works, and could easily overlook it.

I have reported the flaw in this specific context to GitHub, but they've been aware of this behaviour for years and don't consider it a bug, even when abusable for malicious purposes. The report was closed as expected behaviour.  

You can also read the latter part of my blog post: https://cedwards.xyz/github-actions-are-an-impending-security-disaster/

### Describe the solution you'd like

It could be valuable to have an additional pedantic check to ensure that the SHA is from the authentic repository, and not a fork. I can't think of a way to do this without using GItHub's API, which could mean auth tokens are needed, and adds additional friction people don't want to deal with.

**Comment by woodruffw** [comment]
> An example correction for this warning would be to pin the tag to `@3041bf56c941b39c61721a86cd11f3bb1338122a`, ensuring the author of the Action can't 'rugpull' and compromise the workflow by moving the tag to a malicious revision.

This already exists! If you run with `--pedantic` or `--persona=pedantic`, `zizmor` will warn you about unhashed pins:

```
help[unpinned-uses]: unpinned action reference
  --> /Users/william/personal/devel/zizmor/tests/test-data/unpinned-uses.yml:14:9
   |
14 |       - uses: actions/checkout@v3
   |         ------------------------- help: action is not pinned to a hash ref
   |
   = note: audit confidence → High
```

(The reason this is currently a pedantic audit is because it's *very* noisy in practice, and "I trust this org for ref pinning purposes" isn't something that can currently be configured within `zizmor`.)

**Comment by woodruffw** [comment]
> It could be valuable to have an additional pedantic check to ensure that the SHA is from the authentic repository, and not a fork. I can't think of a way to do this without using GItHub's API, which could mean auth tokens are needed, and adds additional friction people don't want to deal with.

FWIW, this is covered by the `impostor-commit` audit (which does indeed require the GitHub API, as you note) 🙂

**Comment by cedws** [comment]
Amazing, thank you.

To use this audit rule, how can I supply a token? Is it via the ordinary `GITHUB_TOKEN` envvar?

I found this post linked in the audit rule's description, this is the first time I came across Chainguard's post about this behaviour too. (https://www.chainguard.dev/unchained/what-the-fork-imposter-commits-in-github-actions-and-ci-cd)

I'm really surprised that GitHub has no interest in resolving this when people are repeatedly highlighting this behaviour as dangerous. They wouldn't even necessarily need to change how objects work - just check that the commits are authentic before executing the workflow.
